torch/_functorch/make_functional.py

```diff
@@ -40,49 +53,57 @@ def _set_nested_attr(obj: nn.Module, names: List[str], value: Tensor) -> None:
         _set_nested_attr(getattr(obj, names[0]), names[1:], value)
 
 
-def _get_nested_attr(obj: nn.Module, names: List[str]) -> None:
+def _get_nested_attr(obj: nn.Module, names: List[str]) -> Union[Tensor, nn.Module]:
     if len(names) == 1:
         return getattr(obj, names[0])
     else:
         return _get_nested_attr(getattr(obj, names[0]), names[1:])
 
 
-def raise_parameter_tying_error():
+def raise_parameter_tying_error() -> NoReturn:
     raise RuntimeError(
         "make_functional(module): we don't yet support models that "
         "do parameter tying (also sometimes known as weight sharing). "
         "Please try to rewrite your model by replacing all instances of the "
         "tied parameter with another and/or comment your support in "
-        "https://github.com/pytorch/functorch/issues/446")
+        "https://github.com/pytorch/functorch/issues/446"
+    )
 
 
-def create_names_map(named_params, tied_named_params):
+def create_names_map(
+    named_params: Union[Dict[str, Tensor], Iterable[Tuple[str, Tensor]]],
+    tied_named_params: Union[Dict[str, Tensor], Iterable[Tuple[str, Tensor]]],
+) -> Dict[str, List[List[str]]]:
     """
     named_params is a dictionary of tensors: {'A': A, 'B': B}
     tied_named_params is another dictionary of tensors {'A': A, 'B': B, 'B_tied': B}
     with potentially tied (or 'duplicated') tensors
 
     This function creates a mapping from the names in named_params to the
-    names in tied_named_params: {'A': ['A'], 'B': ['B', 'B_tied']}.
+    names in tied_named_params: {'A': [['A']], 'B': [['B'], ['B_tied']]}.
     """
-    named_params = {k: v for k, v in named_params}
-    tied_named_params = {k: v for k, v in tied_named_params}
+    named_params = OrderedDict(named_params)
+    tied_named_params = OrderedDict(tied_named_params)
```

Comment:
I don't buy the dict -> OrderedDict change:
- Performance. OrderedDict has a different implementation than dict and there isn't a need to construct a new OrderedDict around a dict here.
- As you mentioned, dict() is already ordered. As of Python 3.7, the language standard says that dict() is ordered. PyTorch does not support Python < 3.7, so there are no BC concerns.
- consistency: I'd like to see buy in for "prefer OrderedDict to dict" in other parts of the PyTorch codebase.

Reply:
> I don't buy the dict -> OrderedDict change:
> 
> * Performance. OrderedDict has a different implementation than dict and there isn't a need to construct a new OrderedDict around a dict here.
> * As you mentioned, dict() is already ordered. As of Python 3.7, the language standard says that dict() is ordered. PyTorch does not support Python < 3.7, so there are no BC concerns.

Let's pass on this since [the language standard guarantees the insertion order for `dict` since Python 3.7](https://docs.python.org/3/library/collections.html#ordereddict-objects). The builtin `dict` is faster for iteration.

```python
In [1]: from collections import OrderedDict

In [2]: iterable = tuple((str(i), i) for i in range(100))

In [3]: %timeit dict(iterable)
2.37 µs ± 20.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [4]: %timeit OrderedDict(iterable)
8.42 µs ± 86.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [5]: d = dict(iterable)

In [6]: od = OrderedDict(iterable)

In [7]: %timeit list(d.items())
2.41 µs ± 83.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [8]: %timeit list(od.items())
4.05 µs ± 87.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

> * consistency: I'd like to see buy in for "prefer OrderedDict to dict" in other parts of the PyTorch codebase.

https://github.com/pytorch/pytorch/blob/3120054c151d37fdf43963dbb60ab420908f48cf/torch/nn/modules/module.py#L448-L463

The attributes in `nn.Module`, such as `_paramters`, `_buffers`, `_modules` are stored as `OrderedDict`.

Note that, `del` and `set` attribute in `_swap_state` changes the order of items in `_paramters` and `_buffers`. (`del + set` is equivalent to `move_to_end()`)

https://github.com/pytorch/pytorch/blob/3120054c151d37fdf43963dbb60ab420908f48cf/torch/_functorch/make_functional.py#L134-L142